tasks/queue-granules/src/index.ts

```diff
@@ -0,0 +1,224 @@
+  return granuleBatch.map((granule) => ({
+    ...granule,
+    createdAt: granule.createdAt ? granule.createdAt : createdAt,
+  }));
+}
+
+/**
+ * See schemas/input.json and schemas/config.json for detailed event description
+ *
+ * @param {HandlerEvent} event - Lambda event object
+ * @returns {Promise<QueueGranulesOutput>} - see schemas/output.json for detailed output schema
+ *   that is passed to the next task in the workflow
+ **/
+async function queueGranules(event: HandlerEvent): Promise<QueueGranulesOutput> {
+  const granules = (event.input.granules || []);
+  const memoizedFetchProvider = memoize(fetchGranuleProvider, (_, providerId) => providerId);
+  const memoizedFetchCollection = memoize(
+    collectionsApi.getCollection,
+    ({ collectionName, collectionVersion }) => constructCollectionId(
+      collectionName,
+      collectionVersion
+    )
+  );
+  const parentExecutionArn = buildExecutionArn(
+    get(event, 'cumulus_config.state_machine')!,
+    get(event, 'cumulus_config.execution_name')!
+  )!;
+  const pMapConcurrency = get(event, 'config.concurrency', 3);
+
+  const messageTemplate = await getJsonS3Object(
+    event.config.internalBucket,
+    templateKey(event.config.stackName)
+  );
+  const { arn: granuleIngestWorkflowArn } = await getJsonS3Object(
+    event.config.internalBucket,
+    getWorkflowFileKey(event.config.stackName, event.config.granuleIngestWorkflow)
+  );
+
+  const executionArns = await pMap(
+    createIterable(granules, event.config.preferredQueueBatchSize),
+    async ({ provider, collectionId, chunks }) => {
+      const { name: collectionName, version: collectionVersion } = deconstructCollectionId(
+        collectionId
+      );
+      const [collection, normalizedProvider] = await Promise.all([
+        memoizedFetchCollection({
+          prefix: event.config.stackName,
+          collectionName,
+          collectionVersion,
+        }),
+        memoizedFetchProvider(event, provider),
+      ]);
+
+      return await pMap(
+        chunks,
+        async (granuleBatchIn) => {
+          const granuleBatch = updateGranuleBatchCreatedAt(granuleBatchIn, Date.now());
+          await pMap(
+            granuleBatch,
+            (queuedGranule) => {
+              const { granuleId, updatedAt, createdAt } = queuedGranule
+
+              if (updatedAt && (!Number.isInteger(updatedAt) || updatedAt < 0)) {
+                throw new Error(`Invalid collection information provided for granule with granuleId: ${granuleId}, please check task input to make sure collection information is provided`)
+              }
+              return granulesApi.updateGranule({
+                prefix: event.config.stackName,
+                collectionId,
+                granuleId,
+                body: {
+                  collectionId,
+                  granuleId,
+                  status: 'queued',
+                  updatedAt: updatedAt!,
+                  createdAt: createdAt!,
```

Comment:
Why do you use the logical not operator for updatedAt and createdAt? 🤔

Reply:
It's the typescript 'non-null assertion operator'. It's not a runtime assertion, just a type assertion.

Reply:
Forgive me for barging in, but the non-null assertion operator should generally be used only when we _know_ a value is non-null, but the compiler is unable to determine that. Using the operator then informs the compiler that we're sure the value is non-null, so don't worry about it (i.e., "Don't worry Compiler. I know exactly what I'm doing here.").

Given the context here, based upon the type information (or lack thereof), we do _not_ know that to be the case. Both (or either) `updatedAt` and `createdAt` could be nullish, so the compiler is indeed correctly identifying this.

In this case, it might be better to resort to using the nullish coalescing operator (`??`) to use a default value when a value is nullish:

```
...,
updatedAt: updatedAt ?? 0,
createdAt: createdAt ?? 0,
...
```

Alternatively, if possible, add more type information to allow the compiler to conclude that they are non-nullish.

In the end, your use of the `!` may not make any material difference here, but the point is that this is arguably not a proper application of the operator, so it may be more of a pedantic issue that I'm pointing out, rather than a practical one. Therefore, feel free to ignore my suggestion.

Reply:
Ah, thank you for the explanation @sbquinlan and @chuckwondo! 🙏  Pretty neat that TypeScript includes a shorthand `!` for non null assertions. 

> Given the context here, based upon the type information (or lack thereof), we do not know that to be the case. Both (or either) updatedAt and createdAt could be nullish, so the compiler is indeed correctly identifying this.

This is a good point, @chuckwondo! However, I _think_? in this case we do know that `createdAt` and `updatedAt` are not nullish since `createdAt` gets set by 
```
const granuleBatch = updateGranuleBatchCreatedAt(granuleBatchIn, Date.now());
```
https://github.com/nasa/cumulus/blob/082e77e65badc708f884ab7c659880296d28234e/tasks/queue-granules/index.js#L144 
and there is a condition to check that `updatedAt` is not nullish
https://github.com/nasa/cumulus/blob/7068fdb6e12b9faad537261215f8da8052b48965/tasks/queue-granules/src/index.ts#L153

Reply:
Ah, nice work following those paths. In that case, it appears that using the non-null assertion operator may indeed be valid here, and thus perhaps simply adding a code comment indicating _how_ you know that it is, since it is not immediately obvious to the reader. A comment simply pointing out the 2 things you just pointed out above should be sufficient to satisfy the reader (such as myself), to avoid such confusion/questioning in the future.

Reply:
https://github.com/nasa/cumulus/pull/3441/commits/b324123d44d80d2bebd8b48903688e4b0bb9b41a#diff-5a524644461453261cb8eed4a96282c50fcd293225634ffdd3ea700eae8d13a8R164-R167

Let me know what you think of this comment, @chuckwondo

Reply:
Actually, upon further review, we do _not_ know that `updatedAt` is non-nullish. Here is the condition you are referring to in the code comment that you added:

```
if (updatedAt && (!Number.isInteger(updatedAt) || updatedAt < 0)) {
  ...
}
```

When `updatedAt` is nullish, it will _bypass_ the `if` block, thus when we reach the line with the expression `updatedAt!`, it will be nullish. Therefore, I do believe that we should be using the nullish coalescing operator instead: `updatedAt ?? 0`, unless you want to do something more involved.

Reply:
Suggested change:

```
                  updatedAt: updatedAt ?? 0,
                  // We use the non-null assertion operator below because we know that
                  // the call above to updateGranuleBatchCreatedAt has set createdAt to
                  // the current time (in milliseconds).
                  createdAt: createdAt!,
  ```

Reply:
https://github.com/nasa/cumulus/pull/3441/commits/1e3e4bb200ddc7026cc51eb228a56814208a0a51#diff-5a524644461453261cb8eed4a96282c50fcd293225634ffdd3ea700eae8d13a8R167